graph/src/components/store.rs

```diff
@@ -397,19 +415,255 @@ pub trait Store: Send + Sync + 'static {
+            .to_owned()
+            .as_string()
+            .ok_or_else(|| format_err!("SubgraphVersion entity has wrong type in `deployment`"))?;
+        SubgraphDeploymentId::new(subgraph_id_str)
+            .map_err(|()| {
+                format_err!("SubgraphVersion entity has invalid subgraph ID in `deployment`")
+            })
+            .map(Some)
+    }
+
+    /// Read all version entities pointing to the specified deployment IDs and determine whether
+    /// they are current or pending in order to produce `SubgraphVersionSummary`s.
+    fn read_subgraph_version_summaries(
+        &self,
+        deployment_ids: Vec<SubgraphDeploymentId>,
+    ) -> Result<(Vec<SubgraphVersionSummary>, Vec<EntityOperation>), Error> {
+        let version_filter = EntityFilter::new_in(
+            "deployment",
+            deployment_ids.iter().map(|id| id.to_string()).collect(),
+        );
+
+        let mut ops = vec![];
+
+        let versions = self.find(SubgraphVersionEntity::query().filter(version_filter.clone()))?;
+        let version_ids = versions
+            .iter()
+            .map(|version_entity| version_entity.id().unwrap())
+            .collect::<Vec<_>>();
+        ops.push(EntityOperation::AbortUnless {
+            description: "Same set of subgraph versions must match filter".to_owned(),
+            query: SubgraphVersionEntity::query().filter(version_filter),
+            entity_ids: version_ids.clone(),
+        });
+
+        // Find subgraphs with one of these versions as current or pending
+        let subgraphs_with_version_as_current_or_pending =
+            self.find(SubgraphEntity::query().filter(EntityFilter::Or(vec![
+                EntityFilter::new_in("currentVersion", version_ids.clone()),
+                EntityFilter::new_in("pendingVersion", version_ids.clone()),
+            ])))?;
+        let subgraph_ids_with_version_as_current_or_pending =
+            subgraphs_with_version_as_current_or_pending
+                .iter()
+                .map(|subgraph_entity| subgraph_entity.id().unwrap())
+                .collect::<HashSet<_>>();
+        ops.push(EntityOperation::AbortUnless {
+            description: "Same set of subgraphs must have these versions as current or pending"
+                .to_owned(),
+            query: SubgraphEntity::query().filter(EntityFilter::Or(vec![
+                EntityFilter::new_in("currentVersion", version_ids.clone()),
+                EntityFilter::new_in("pendingVersion", version_ids),
+            ])),
+            entity_ids: subgraph_ids_with_version_as_current_or_pending
+                .into_iter()
+                .collect(),
+        });
+
+        // Produce summaries, deriving flags from information in subgraph entities
+        let version_summaries =
+            versions
+                .into_iter()
+                .map(|version_entity| {
+                    let is_current = subgraphs_with_version_as_current_or_pending.iter().any(
```

Comment:
Is there anything that guarantees that a subgraph version always belongs to only one subgraph? I know the updating logic probably prevents that but just looking at this code you could think that a version could be considered  current or pending if it is current or pending for one subgraph but not for others.

It makes sense to batch the processing  here to keep it simple and I know this logic is correct but I wonder if we can make it more obvious here.

Reply:
Ah, I'll add some asserts! It is guaranteed that a subgraph version points to a single subgraph because the subgraph version entity points to only one subgraph, but it's only implicit that Subgraph.currentVersion.subgraph_id matches Subgraph.id